src/lib/insights/client.ts

```diff
@@ -0,0 +1,99 @@
+import {
+  Hits,
+  InsightsClient,
+  InsightsClientMethod,
+  InsightsClientPayload,
+  Renderer,
+  RenderOptions,
+  Unmounter,
+  WidgetFactory,
+} from '../../types';
+import find from 'lodash/find';
+
+export const inferPayload = (
+  method: InsightsClientMethod,
+  results: RenderOptions['results'],
+  hits: Hits<unknown>,
+  objectIDs: string[]
+): InsightsClientPayload => {
+  const { index, queryID } = results;
+  switch (method) {
+    case 'clickedObjectIDsAfterSearch': {
+      const positions = objectIDs.map(objectID => {
+        const hit = find(hits, hit => hit.objectID === objectID);
+        // NOTE: we might want to optimize this so as we browse though the
+        // hits once, which can get pretty big
+        if (typeof hit === 'undefined') {
+          throw new Error(
+            `Could not find ObjectID ${objectID} in the returned hits.`
+          );
+        }
+        return hit.__position;
+      });
+      return { index, queryID, objectIDs, positions };
+    }
+
+    case 'convertedObjectIDsAfterSearch':
+      return { index, queryID, objectIDs };
+
+    default:
+      throw new Error(`Unsupported method passed ${method}`);
+  }
+};
+
+const _wrapInsightsClient = (
+  aa: InsightsClient,
+  results: RenderOptions['results'],
+  hits: Hits<unknown>
+): InsightsClient => (
+  method: InsightsClientMethod,
+  payload: InsightsClientPayload
+) => {
+  if (typeof aa !== 'function') {
+    throw new TypeError(`Expected insightsClient to be a Function`);
+  }
+  if (!Array.isArray(payload.objectIDs)) {
+    throw new TypeError(`Expected objectIDs to be an Array`);
+  }
+  const inferredPayload = inferPayload(
+    method,
+    results,
+    hits,
+    payload.objectIDs
+  );
+  aa(method, { ...inferredPayload, ...payload });
+};
+
+type Connector<WidgetParams> = (
+  renderFn: Renderer<RenderOptions<WidgetParams>>,
+  unmountFn: Unmounter
+) => WidgetFactory<WidgetParams>;
+
+export default function withInsights(
+  connector: Connector<unknown>
+): Connector<unknown> {
+  const _wrapRenderFn = (
+    renderFn: Renderer<RenderOptions<unknown>>
+  ): Renderer<RenderOptions<unknown>> => (
+    renderOptions: RenderOptions,
+    isFirstRendering: boolean
```

Comment:
Recently we took the initiative to rename this by `isFirstRender` (that's the new convention).